### PR TITLE
implemented javascript runtime

### PR DESCRIPTION
**Pull request by darmie**
It embeds a [rquickjs](https://github.com/DelSkayn/rquickjs) javascript runtime in a zflow component and evaluates an ES5 or ES6 standard javascript. ProcessHandle object is added as a global variable.
```js
const inport = ProcessHandle.inport;
const outport = ProcessHandle.outport;


if ("left" in inport && "right" in inport) {
    if(inport.left && inport.right){
        outport.send({sum: Number(inport.left) + Number(inport.right)})
    }
}
```

It does not support module and value exports yet.

### CODE REVIEW

File: zflow_runtime/src/js/mod.rs [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    pub metadata: Map<String, Value>,
+}
+
+impl JsComponent {
+    pub fn from_metadata(meta: Value) -> Option<JsComponent> {
+        JsComponent::deserialize(meta).ok()
+    }
+
+    pub fn with_metadata(&mut self, meta: Value) -> JsComponent {
+        if let Some(meta) = JsComponent::deserialize(meta.clone()).ok() {
```
Comment (pyjac) [inline zflow_runtime/src/js/mod.rs:71]: `JsComponent::from_metadata` could be used instead of JsComponent::deserialize(meta.clone()).ok(), since it does the same thing

File: zflow_runtime/src/js/mod.rs [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+            }
+            self.forward_brackets.extend(meta.forward_brackets);
+            if !meta.base_dir.is_empty() {
+                self.base_dir = meta.base_dir;
+            }
+        } else if let Some(meta) = meta.clone().as_object() {
+            self.metadata = meta.clone();
+        }
+
+        self.clone()
```
Comment (pyjac) [inline zflow_runtime/src/js/mod.rs:88]: no sure i understand why you need a clone here... i assume this object, "itself" would be used somewhere after this merge with the passed meta

File: zflow_runtime/src/js/mod.rs [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    }
+}
+
+// lazy_static!{
```
Comment (pyjac) [inline zflow_runtime/src/js/mod.rs:92]: unused?

File: zflow_runtime/src/js/mod.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+//     pub static ref JS_RUNTIME:Runtime = {
+//         let rt = Rc::new(RefCell::new(Runtime::new().unwrap()));
+//         // Todo: set resolver and loader for inbuilt modules
+//         // rt.set_loader(resolver, loader);
+//         rt
+//     };
+// }
+
+impl ModuleComponent for JsComponent {
+    fn as_component(&self) -> Result<Component, String> {
```
Comment (pyjac) [inline zflow_runtime/src/js/mod.rs:102]: i see lots of cloning going here.. is there a way to somehow review if the values can be passed as ref?
also building ComponentOptions is quite some series of actions that's a little hard to follow in a single  and long method, maybe break into methods

File: zflow_runtime/Cargo.toml [hunk 5]
```diff
@@ -31,6 +31,8 @@ extism = "0.3.0"
 wasmtime = "6.0.2"
 anyhow = "1.0.71"
 bincode = "1.3.3"
+rquickjs = {version = "0.1.7", features = ["exports", "async-std", "rust-alloc", "loader", "dyn-load"]}
+lazy_static = "1.4.0"
```
[the remark this answers is not in the document]
Comment (darmie) [reply zflow_runtime/Cargo.toml:35]: What wasn't used? lazy_static?

[the remark this answers is not in the document]
Comment (pyjac) [reply zflow_runtime/Cargo.toml:35]: yea. lazy_static
